Handle op outputs in default asset IO manager

clairelin135 opened the pull request:
Addresses https://github.com/dagster-io/dagster/issues/7713

This PR adjusts the default `fs_asset_io_manager` to handle op outputs by storing op outputs in a file directory `/run_id/step_key` to ensure unique directories for each step.

OwenKephart commented on python_modules/dagster/dagster/core/storage/fs_asset_io_manager.py:
```diff
@@ -81,4 +81,8 @@ def asset2(asset1):
+        if context.has_asset_key:
+            path = context.get_asset_output_identifier()
+        else:
+            path = [context.run_id, context.step_key]
```
nit: can this be context.get_run_scoped_asset_identifier?

clairelin135 replied:
Thanks! didn't realize that function existed

OwenKephart reviewed (approved):
one comment, otherwise 🚢

sryza commented:
If I understand correctly, this makes the behavior of fs_asset_io_manager identical to the behavior of fs_io_manager, when the output is not an asset. Thoughts on just merging them together? I.e. getting rid of fs_asset_io_manager and adding logic inside fs_io_manager?  A small advantage is that we'd now have a name that's not as long.

Also, we should merge this after this external contribution lands so that we don't force them to rebase: https://github.com/dagster-io/dagster/pull/8007.

sryza reviewed (changes requested):
.

clairelin135 commented on python_modules/dagster/dagster/core/asset_defs/assets_job.py:
```diff
@@ -131,7 +131,7 @@ def asset2(asset1):
     return graph.to_job(
         resource_defs=merge_dicts(
-            {"io_manager": fs_asset_io_manager}, all_resource_defs, {"root_manager": root_manager}
+            {"io_manager": fs_io_manager}, all_resource_defs, {"root_manager": root_manager}
```
removed ✅

clairelin135 commented:
@sryza I've adjusted this PR to merge functionality of `fs_asset_io_manager` with `fs_io_manager` (and deleted references to `fs_asset_io_manager`). I think this is ready for you to take another look!

sryza reviewed (approved):
Nice! Love to see PRs that remove more lines than they add. We'll need to make sure to message this as a breaking change in our CHANGES.md. Also, as followups, we'll need this for the gcs, s3, and azure IO managers as well.
